### ISSUE TITLE
Allow MediaSource instances to be reused

### ISSUE THREAD

**Issue by stoleruedgar** [google/ExoPlayer#3498]
### Issue description
Passing the same DynamicConcatenatingMediaSource to a new SimpleExoPlayer instance doubles the internal mediaSourceHolders which in turn breaks the Player behaviour

### Reproduction steps
```
public class TestActivity extends Activity {

	private DynamicConcatenatingMediaSource playlist = null
	private SimpleExoPlayer player = null

	protected onCreate(Bundle savedInstance) {
		super.onCreate(savedInstance)
		playlist = ....init with mediasources       
	}	

	protected onStart() {
		super.onStart()      
		if (player == null) {
			player = ...init player
			if (playlist != null) {
				player.prepare(playlist, false, true)
			}
		} 
	}

	protected onStop() {
		super.onStop()
		player = ...release player
	}	
}
```
Put the activity in background and then bring the app in the foreground.

### Version of ExoPlayer being used
r2.5.4

### Device(s) and version(s) of Android being used
Samsung Galaxy S7, Android 7.0

**Comment by tonihei** [comment]
Media sources shouldn't be reused ([documented here](http://google.github.io/ExoPlayer/doc/reference/com/google/android/exoplayer2/ExoPlayer.html#prepare-com.google.android.exoplayer2.source.MediaSource-)). This includes the dynamic concatenating media source as well as the single playlist item media sources.

**Comment by ojw28** [comment]
Could/should we add some assertions that check this and give a developer friendly error message? Seems like an easy mistake to make.

**Comment by tonihei** [comment]
I think we can also make DynamicConcatenatingMediaSource reusable (despite the general guideline). This would deliberately allow reusing the wrapping dynamic media source. Marking this as enhancement to do this.

**Comment by tonihei** [comment]
To clarify: this enhancement tracks reusing media sources in general. Reusing just the DynamicConcatenatingMediaSource wouldn't make sense if the contained media sources can't be reused.

**Comment by eneim** [comment]
@tonihei @ojw28 it is great to see that MediaSource can be reused. I still not see javadoc update for it in dev-v2. Do you plan it for 2.7.0 or it must be in the release after?

I have a couple question about this:

- What I need to do to *reset* the MediaSource so that it can be reuse right after. (with the latest update in dev-v2)
- Is ``ExoPlayer#prepare(MediaSource)`` an expensive operation? Right now I'm trying to reuse MediaSource (2.6.1) by cloning current MediaSource and prepare it again. It sounds weird, but have no other way to do ...

Thanks.

**Comment by tonihei** [comment]
The current updates mentioned above are only one part of making MediaSource reusable. You should now be able to re-prepare with the same media source. However, we are still working on the possibility to reuse media source in parallel, e.g. adding them twice to a ConcatenatingMediaSource. That's also why we haven't updated the docs yet because we only do this after the feature is fully implemented and ready to use. It will most likely be part of 2.7.0.

For the second question: `prepare` is not in itself an expensive operation but it might trigger other operations. For example, manifest files for DASH, SmoothStreaming or HLS are downloaded after calling `prepare`.

**Comment by tonihei** [comment]
Closing this issue as the feature is now available in the dev-branch.

**Comment by prat-z** [comment]
@tonihei @ojw28 Will the reuse of MediaSource enable a MediaSource object to be shared between more than one players, like one player in a list and one full screen, playing the same media  ?
And, when will this change be available on the release branch?

**Comment by tonihei** [comment]
@prat-z Sorry, this will be the only restriction. Using the same media source simultaneously from multiple players would require us to synchronize the access to the source which we currently avoid by using it from one player at a time only. However, creating a media source isn't that cost-intensive and this is really mostly a convenience feature. So you wouldn't save much by avoiding to create a new media source for your second player anyway. 

This feature is already in the dev-branch if you want to try it out. And it's going to be part of release 2.8.

**Comment by prat-z** [comment]
@tonihei The problem with creating a new media source is that I am using HLSMediaSource, and creating a new data source means loosing out on the downloaded bytes of the same video, and a buffering state while switching to and from full screen mode.

**Comment by tonihei** [comment]
@prat-z Reusing media source wouldn't solve that I'm afraid. To keep the buffered video data when switching to and from full screen mode, you need to make sure to keep the same player instance. Please have a look around our issue tracker for fullscreen related issues. There are also various suggestions on the web how to implement fullscreen switching with ExoPlayer. Because of the different behavior on different Android versions, there is unfortunately no single, simple way to do this. 

If you have more question regarding full-screen switching, please open a new issue.
